src/Ocean/Domains/rectangular_domain.jl

```diff
@@ -0,0 +1,194 @@
+
+Returns a `RectangularDomain` representing the product of `x, y, z` intervals,
+specified by 2-tuples.
+
+The `RectangularDomain` is meshed with a simple `DiscontinuousSpectralElementGrid`
+with an isotropic `polynomialorder` and a 3-tuple of `elements`
+giving the number of elements in `x, y, z`.
+
+Additional arguments are:
+
+- `periodicity`: a 3-tuple that indicates periodic dimensions with `true`, 
+
+- `boundary`: specifies the boundary condition on each boundary with a
+              boundary condition `tag`
+
+- `array_type`: either `Array` for CPU computations or `CuArray` for
+                GPU computations
+
+- `message_communicator`: the world communicator for message passing between
+                          nodes in a distributed memory configuration.
+
+Example
+=======
+
+```jldoctest
+julia> using ClimateMachine; ClimateMachine.init()
+
+julia> using ClimateMachine.Ocean.RectangularDomains: RectangularDomain
+
+julia> domain = RectangularDomain(elements=(7, 8, 9), polynomialorder=4, x=(0, 1), y=(0, 1), z=(0, 1))
+RectangularDomain{Float64, ClimateMachine.Mesh.Grids.DiscontinuousSpectralElementGrid}:
+    Np = 4, Ne = (x = 7, y = 8, z = 9)
+    L = (x = 1.00e+00, y = 1.00e+00, z = 1.00e+00)
+    x = (0.00e+00, 1.00e+00), y = (0.00e+00, 1.00e+00), z = (1.00e+00, 0.00e+00)
+```
+"""
+function RectangularDomain(
+    FT = Float64;
+    elements,
+    polynomialorder,
+    x::Tuple{<:Number, <:Number},
+    y::Tuple{<:Number, <:Number},
+    z::Tuple{<:Number, <:Number},
+    periodicity = (true, true, false),
+    boundary = ((0, 0), (0, 0), (1, 2)),
+    array_type = Array,
+    message_communicator = MPI.COMM_WORLD,
+)
+
+    Ne = name_it(elements)
+
+    west, east = FT.(x)
+    south, north = FT.(y)
+    bottom, top = FT.(z)
+
+    east > west || error("Domain x-limits must be increasing!")
+    north > south || error("Domain y-limits must be increasing!")
+    top > bottom || error("Domain z-limits must be increasing!")
+
+    L = (x = east - west, y = north - south, z = top - bottom)
+
+    element_coordinates = (
+        range(west, east, length = Ne.x + 1),
+        range(south, north, length = Ne.y + 1),
+        range(bottom, top, length = Ne.z + 1),
+    )
+
+    topology = StackedBrickTopology(
+        message_communicator,
+        element_coordinates;
+        periodicity = periodicity,
+        boundary = boundary,
+    )
+
+    grid = DiscontinuousSpectralElementGrid(
+        topology,
+        FloatType = FT,
+        DeviceArray = array_type,
+        polynomialorder = polynomialorder,
+    )
+
+    return RectangularDomain{FT, typeof(grid)}(
+        grid,
+        polynomialorder,
+        Ne,
+        L,
+        (west, east),
+        (south, north),
+        (bottom, top),
+    )
+end
+
+array_type(domain::RectangularDomain) = Array #array_type(domain.grid)
+eltype(::RectangularDomain{FT}) where {FT} = FT
+communicator(args...) = MPI.COMM_WORLD
```

Comment:
This looks like it's grabbing `MPI` globally--we probably don't want to do that. cc @kpamnany

Reply:
For sure, that's why I put this place holder function here. Please advise.

Reply:
We use the name `mpicomm` everywhere. Not sure what this `communicator` call is used for... the communicator is ideally passed into the constructor for internal use by the module.

Reply:
Is "mpicomm" an abbreviation for "message_communicator"? Is there a function "mpicomm" that I can use?

As a student, I was eternally confused and frustrated reading FORTRAN codes and running into the name "MPI_COMM_WORLD" (or whatever) and having no idea what it meant or what it did. It's short for something --- but what?!?! It seems important but is so inscrutable. Bundle that experience with more general anxiety about learning new things and it becomes a toxic mess.

Now that I know that it literally "communicates messages" I feel `message_communicator` is the right name. I'm trying to help future Greg's...

Reply:
The reason why it's needed here is because `DiscontinuousSpectralElementGrid` needs a communicator. But the grid is built before the model, so to use the same communicator in the model and the grid, I need to query the grid to get its communicator.

Rather than hard code a reference to the communicator via domain/grid, I wrote this function, anticipating that the design would change shortly...

I added a naive, simple default that would work in the cases I knew I would use it in, with the plan of extending / changing it in the future when I had examples of different communicators being used.

I'm happy to change this design.

Reply:
> Is "mpicomm" an abbreviation for "message_communicator"? Is there a function "mpicomm" that I can use?

It means "MPI communicator". It isn't a function -- I meant that the name of the variable used to hold an MPI communicator (in all of the existing ClimateMachine code) is `mpicomm`.

We could use something besides MPI to do HPC ([GASNet](https://gasnet.lbl.gov/), [Global Arrays](https://hpc.pnl.gov//globalarrays/), etc.) but we're not -- this is an MPI code.
> 
> As a student, I was eternally confused and frustrated reading FORTRAN codes and running into the name "MPI_COMM_WORLD" (or whatever) and having no idea what it meant or what it did. It's short for something --- but what?!?! It seems important but is so inscrutable. Bundle that experience with more general anxiety about learning new things and it becomes a toxic mess.

Take deep breaths! Then read at least the beginning of [this](https://pages.tacc.utexas.edu/~eijkhout/pcse/html/mpi-comm.html).

> Now that I know that it literally "communicates messages" I feel `message_communicator` is the right name. I'm trying to help future Greg's...

I think `communicator` is fine for this -- see next comment.

Reply:
> The reason why it's needed here is because `DiscontinuousSpectralElementGrid` needs a communicator. But the grid is built before the model, so to use the same communicator in the model and the grid, I need to query the grid to get its communicator.

Makes excellent sense. I suggest the implementation be:
```julia
communicator(domain::RectangularDomain) = communicator(domain.grid)
```
And add:
```julia
communicator(grid::DiscontinuousSpectralElementGrid) = communicator(grid.topology)
```
Somewhere [here](https://github.com/CliMA/ClimateMachine.jl/blob/ocean/glw/super-hydrostatic-boussinesq/src/Numerics/Mesh/Grids.jl#L277), and add:
```julia
communicator(topology::BoxElementTopology) = topology.mpicomm
```
Somewhere [here](https://github.com/CliMA/ClimateMachine.jl/blob/ocean/glw/super-hydrostatic-boussinesq/src/Numerics/Mesh/Grids.jl#L277) as well as a few other similar definitions for the hierarchy of topologies (primarily [these](https://github.com/CliMA/ClimateMachine.jl/blob/master/src/Numerics/Mesh/Topologies.jl#L7-L10) I think).

Reply:
My original comment should have been targeted at [this line](https://github.com/CliMA/ClimateMachine.jl/blob/ocean/glw/super-hydrostatic-boussinesq/src/Ocean/Domains/rectangular_domain.jl#L146) -- please use `mpicomm` for the argument name there for consistency with the rest of the code.

Reply:
Cc: @jkozdon for any thoughts.

Reply:
Great, thanks. I'm certainly not suggesting to use something other than MPI, I hope I didn't imply that. I was really just pointing out that long ago this was a confusion for me. I'm using my past experiences as a guide for how other, similar people might feel when reading my code.

Looking into the code it looks like `mpicomm` is a property of `grid.topology`... ? I didn't know exactly where it was before, hence the fallback / function. Is `grid.topology` reliable for inferring the `communicator`? How would a user specify a different `communicator` if they needed to?

Reply:
Ok, I will use the name `mpicomm` for the MPI communicator.

Reply:
It's probably not appropriate for me to put these deep changes into this PR, which is rather experimental. Can we do something simple temporarily in order to make progress quickly, and open an issue about writing an infrastructure for using consistent communicators? I'm happy to implement the PR that resolves the issue too, I just think this PR is growing, and growing...

Reply:
No, it's a reasonable concern, hence I've pulled Jeremy in to comment. Previously, we were using the global communicator exclusively (your favorite `MPI_COMM_WORLD`). Then, [this](https://github.com/CliMA/ClimateMachine.jl/blob/ocean/glw/super-hydrostatic-boussinesq/src/Arrays/MPIStateArrays.jl#L128-L131) was added.

Wait... when you say:
> The reason why it's needed here is because DiscontinuousSpectralElementGrid needs a communicator. But the grid is built before the model, so to use the same communicator in the model and the grid, I need to query the grid to get its communicator.

What's the model now? The `*Equations` shouldn't need a communicator, and the `SuperModel` is constructing the grid, so it should already have the communicator...

Reply:
@jkozdon @kpamnany @charleskawczynski I'm resolving this issue by removing the `communicator` function and using COMM_WORLD when building DriverConfiguration (wihch doesnt matter becuaase we will nuke DriverConfiguration anyways). We can discuss communicator ju-jitsu when the time comes!